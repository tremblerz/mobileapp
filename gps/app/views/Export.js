import PropTypes from 'prop-types';
import React, { useEffect } from 'react';
import { useTranslation } from 'react-i18next';
import {
  BackHandler,
  NativeModules,
  SafeAreaView,
  ScrollView,
  StatusBar,
  StyleSheet,
  View,
} from 'react-native';
import RNFS from 'react-native-fs';
import LinearGradient from 'react-native-linear-gradient';
import Share from 'react-native-share';
import RNFetchBlob from 'rn-fetch-blob';

import close from '../../../shared/assets/svgs/close';
import exportIcon from '../../../shared/assets/svgs/export';
import { isPlatformiOS } from './../Util';
import { Button } from '../components/Button';
import { IconButton } from '../components/IconButton';
import { Typography } from '../components/Typography';
import Colors from '../constants/colors';
import fontFamily from '../constants/fonts';
import { Theme } from '../constants/themes';

const base64 = RNFetchBlob.base64;

export const ExportScreen = ({ navigation }) => {
  const { t } = useTranslation();
  function handleBackPress() {
    navigation.goBack();
    return true;
  }

  useEffect(() => {
    BackHandler.addEventListener('hardwareBackPress', handleBackPress);

    return function cleanup() {
      BackHandler.removeEventListener('hardwareBackPress', handleBackPress);
    };
  });

  function backToMain() {
    navigation.goBack();
  }

  async function onShare() {
    try {
      let locationData = await NativeModules.SecureStorageManager.getLocations();
      let nowUTC = new Date().toISOString();
      let unixtimeUTC = Date.parse(nowUTC);

      let options = {};
      let jsonData = JSON.stringify(locationData);
      const title = 'COVIDSafePaths.json';
      const filename = unixtimeUTC + '.json';
      const message = 'Here is my location log from COVID Safe Paths.';
      if (isPlatformiOS()) {
        const url = RNFS.DocumentDirectoryPath + '/' + filename;
        await RNFS.writeFile(url, jsonData, 'utf8')
          .then(() => {
            options = {
              activityItemSources: [
                {
                  placeholderItem: { type: 'url', content: url },
                  item: {
                    default: { type: 'url', content: url },
                  },
                  subject: {
                    default: title,
                  },
                  linkMetadata: { originalUrl: url, url, title },
                },
              ],
            };
          })
          .catch(err => {
            console.log(err.message);
          });
      } else {
        jsonData = 'data:application/json;base64,' + base64.encode(jsonData);
        options = {
          title,
          subject: title,
          url: jsonData,
          message: message,
          filename: filename,
        };
      }
      await Share.open(options)
        .then(res => {
          console.log(res);
        })
        .catch(err => {
          console.log(err);
          console.log(err.message, err.code);
        });
      if (isPlatformiOS()) {
        // eslint-disable-next-line no-undef
        await RNFS.unlink(url);
      }
    } catch (error) {
      console.log(error.message);
    }
  }

  return (
    <Theme use='violet'>
      <StatusBar
        barStyle='light-content'
        backgroundColor={Colors.VIOLET_BUTTON}
        translucent={false}
      />
      <SafeAreaView style={styles.topSafeAreaContainer} />
      <SafeAreaView style={styles.bottomSafeAreaContainer}>
        <LinearGradient
          start={{ x: 0, y: 0 }}
          end={{ x: 0, y: 1 }}
          colors={[Colors.VIOLET_BUTTON, Colors.VIOLET_BUTTON_DARK]}
          style={{ flex: 1, height: '100%' }}>
          <View style={styles.headerContainer}>
            <IconButton
              icon={close}
              size={18}
              onPress={() => backToMain()}
              accessibilityLabel='Close'
            />
          </View>

          <ScrollView contentContainerStyle={styles.contentContainer}>
            <View style={styles.main}>
              <Typography use='headline2' style={styles.exportSectionTitles}>
                {t('share.title')}
              </Typography>
              <Typography use='body1' style={styles.exportSectionPara}>
                {t('share.paragraph_first')}
              </Typography>
              <Typography use='body1' style={styles.exportSectionPara}>
                {t('share.paragraph_second')}
              </Typography>

              <Button
                style={styles.exportButton}
                label={t('share.button_text')}
                icon={exportIcon}
                onPress={onShare}
              />
            </View>
          </ScrollView>
        </LinearGradient>
      </SafeAreaView>
    </Theme>
  );
};

const styles = StyleSheet.create({
  // Container covers the entire screen
  topSafeAreaContainer: {
    flex: 0,
    backgroundColor: Colors.VIOLET_BUTTON,
  },
  bottomSafeAreaContainer: {
    flex: 1,
    backgroundColor: Colors.VIOLET_BUTTON_DARK,
  },
  headerContainer: {
    alignItems: 'center',
    flexDirection: 'row',
    height: 55,
    justifyContent: 'flex-end',
    paddingHorizontal: 26,
  },
  contentContainer: {
    flexDirection: 'column',
    width: '100%',
    flex: 1,
    paddingHorizontal: 26,
  },
  exportSectionTitles: {
    marginTop: 9,
    fontWeight: 'normal',
    fontFamily: fontFamily.primaryMedium,
  },
  exportSectionPara: {
    marginTop: 22,
  },
  exportButton: {
    marginTop: 48,
  },
  main: {
    flex: 1,
    paddingTop: 48,
  },
});

ExportScreen.propTypes = {
  shareButtonDisabled: PropTypes.bool,
};

ExportScreen.defaultProps = {
  shareButtonDisabled: true,
};
