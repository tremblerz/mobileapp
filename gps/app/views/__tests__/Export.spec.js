import { render } from '@testing-library/react-native';
import React from 'react';

import { ExportScreen } from '../Export';

describe('<ExportScreen />', () => {
  it('renders correctly', () => {
    const { asJSON } = render(<ExportScreen />);
    expect(asJSON()).toMatchSnapshot();
  });
});
