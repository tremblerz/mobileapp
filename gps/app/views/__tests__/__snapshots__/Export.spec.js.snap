// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ExportScreen /> renders correctly 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Object {
        "backgroundColor": "#6979F8",
        "flex": 0,
      }
    }
  />
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Object {
        "backgroundColor": "#3A4CD7",
        "flex": 1,
      }
    }
  >
    <BVLinearGradient
      colors={
        Array [
          4285102584,
          4282010839,
        ]
      }
      endPoint={
        Object {
          "x": 0,
          "y": 1,
        }
      }
      locations={null}
      startPoint={
        Object {
          "x": 0,
          "y": 0,
        }
      }
      style={
        Object {
          "flex": 1,
          "height": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "height": 55,
            "justifyContent": "flex-end",
            "paddingHorizontal": 26,
          }
        }
      >
        <View
          accessibilityLabel="Close"
          accessible={true}
          focusable={true}
          isTVSelectable={true}
          style={
            Object {
              "alignContent": "center",
              "alignItems": "center",
              "opacity": 1,
            }
          }
        >
          <SvgXml
            height={18}
            style={
              Object {
                "color": "#FFF",
              }
            }
            width={18}
            xml="
<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<svg width=\\"24\\" height=\\"24\\" viewBox=\\"0 0 24 24\\" fill=\\"none\\" xmlns=\\"http://www.w3.org/2000/svg\\">
  <path d=\\"M3.25731 22.5795L3.25737 22.5796L3.26062 22.5762L11.9494 13.3715L20.6888 22.6609L20.6887 22.661L20.691 22.6632C20.9628 22.9377 21.2962 23.125 21.789 23.125C22.4855 23.125 23.125 22.6697 23.125 22.0232C23.125 21.6762 22.9451 21.4335 22.7374 21.215L13.7371 11.7456L22.3837 2.74296L22.3842 2.74249C22.6022 2.51308 22.7213 2.27133 22.7213 1.97683C22.7213 1.38538 22.1473 0.875 21.4862 0.875C20.9837 0.875 20.6509 1.11422 20.339 1.42046L20.339 1.42041L20.3359 1.42364L12.0506 10.1618L3.71471 1.33886L3.71474 1.33884L3.71268 1.33676C3.4409 1.06226 3.10745 0.875 2.61468 0.875C1.91817 0.875 1.27867 1.33027 1.27867 1.97683C1.27867 2.32383 1.45862 2.56647 1.66628 2.78496L1.66648 2.78517L10.2629 11.7876L1.21239 21.2573L1.21215 21.2575C0.994108 21.4869 0.875 21.7287 0.875 22.0232C0.875 22.6146 1.44905 23.125 2.11009 23.125C2.61264 23.125 2.94544 22.8858 3.25731 22.5795Z\\" fill=\\"currentColor\\" stroke=\\"currentColor\\" stroke-width=\\"0.25\\"/>
</svg>
"
          />
        </View>
      </View>
      <RCTScrollView
        contentContainerStyle={
          Object {
            "flex": 1,
            "flexDirection": "column",
            "paddingHorizontal": 26,
            "width": "100%",
          }
        }
      >
        <View>
          <View
            style={
              Object {
                "flex": 1,
                "paddingTop": 48,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#FFF",
                  "fontFamily": "IBMPlexSans-Medium",
                  "fontSize": 26,
                  "fontWeight": "normal",
                  "lineHeight": 34,
                  "marginTop": 9,
                  "writingDirection": "ltr",
                }
              }
              use="headline2"
            >
              Share location history
            </Text>
            <Text
              style={
                Object {
                  "color": "#FFF",
                  "fontFamily": "IBMPlexSans",
                  "fontSize": 18,
                  "fontWeight": "normal",
                  "lineHeight": 24,
                  "marginTop": 22,
                  "writingDirection": "ltr",
                }
              }
              use="body1"
            >
              If you test positive for COVID-19, please do your part by sharing your location history with local authorities.
            </Text>
            <Text
              style={
                Object {
                  "color": "#FFF",
                  "fontFamily": "IBMPlexSans",
                  "fontSize": 18,
                  "fontWeight": "normal",
                  "lineHeight": 24,
                  "marginTop": 22,
                  "writingDirection": "ltr",
                }
              }
              use="body1"
            >
              Location is shared as a simple list of times and places, no additional information.
            </Text>
            <View
              accessibilityLabel="Share location data"
              accessibilityRole="button"
              accessible={true}
              focusable={true}
              isTVSelectable={true}
              style={
                Object {
                  "alignContent": "center",
                  "alignItems": "center",
                  "alignSelf": "stretch",
                  "backgroundColor": "#FFF",
                  "borderBottomLeftRadius": 8,
                  "borderBottomRightRadius": 8,
                  "borderColor": "transparent",
                  "borderStyle": "solid",
                  "borderTopLeftRadius": 8,
                  "borderTopRightRadius": 8,
                  "borderWidth": 2,
                  "flexDirection": "row",
                  "height": 72,
                  "justifyContent": "space-between",
                  "marginTop": 48,
                  "opacity": 1,
                  "paddingHorizontal": 16,
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "#4051DB",
                    "fontFamily": "IBMPlexSans-Medium",
                    "fontSize": 20,
                    "fontWeight": "normal",
                    "lineHeight": 40,
                    "writingDirection": "ltr",
                  }
                }
                use="body1"
              >
                Share location data
              </Text>
              <SvgXml
                style={
                  Object {
                    "marginRight": 8,
                  }
                }
                xml="<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<svg width=\\"16\\" height=\\"21\\" viewBox=\\"0 0 16 21\\" fill=\\"none\\" xmlns=\\"http://www.w3.org/2000/svg\\">
<path d=\\"M13 7H14C15.1046 7 16 7.89543 16 9V19C16 20.1046 15.1046 21 14 21H2C0.89543 21 0 20.1046 0 19V9C0 7.89543 0.89543 7 2 7H3C3.55228 7 4 7.44772 4 8C4 8.55228 3.55228 9 3 9H2V19H14V9H13C12.4477 9 12 8.55228 12 8C12 7.44772 12.4477 7 13 7ZM7.24029 0.349711C7.42369 0.135645 7.69601 0 8 0C8.30399 0 8.57631 0.135645 8.75971 0.349711L11.5355 3.12553C11.9261 3.51606 11.9261 4.14922 11.5355 4.53975C11.145 4.93027 10.5118 4.93027 10.1213 4.53975L9 3.41843V13C9 13.5523 8.55228 14 8 14C7.44772 14 7 13.5523 7 13V3.41843L5.87868 4.53975C5.48815 4.93027 4.85499 4.93027 4.46447 4.53975C4.07394 4.14922 4.07394 3.51606 4.46447 3.12553L7.24029 0.349711Z\\" fill=\\"#4051DB\\"/>
</svg>"
              />
            </View>
          </View>
        </View>
      </RCTScrollView>
    </BVLinearGradient>
  </RCTSafeAreaView>
</View>
`;
